aten/src/ATen/native/Blas.cpp

```diff
@@ -4,60 +4,35 @@
-scalar_t vdot_impl(int64_t n, scalar_t *x, int64_t incx, scalar_t *y, int64_t incy);
-
-constexpr inline bool lda_cond(int64_t m, int64_t n, int64_t lda) {
-  return n == 1 || lda >= std::max<int64_t>(1L, m);
-}
-
-Tensor &addmv_impl_cpu(Tensor& result, const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta_, const Scalar& alpha_) {
-  auto r_stride = result.stride(0);
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16, mat.scalar_type(), "addmv_impl_cpu", [&] {
-    auto beta = beta_.to<scalar_t>();
-    auto alpha = alpha_.to<scalar_t>();
-    if (mat.stride(0) == 1 && lda_cond(mat.size(0), mat.size(1), mat.stride(1))) {
-      gemv<scalar_t>('n', mat.size(0), mat.size(1), alpha, mat.data_ptr<scalar_t>(), mat.stride(1),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-    else if (mat.stride(1) == 1 && lda_cond(mat.size(1), mat.size(0), mat.stride(0))) {
-      gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, mat.data_ptr<scalar_t>(), mat.stride(0),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-    else {
-      Tensor cmat = mat.contiguous();
-      gemv<scalar_t>('t', mat.size(1), mat.size(0), alpha, cmat.data_ptr<scalar_t>(), cmat.stride(0),
-          vec.data_ptr<scalar_t>(), vec.stride(0), beta, result.data_ptr<scalar_t>(), r_stride);
-    }
-  });
-  return result;
-}
-
-Tensor &addmv_out(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha, Tensor& result) {
-  { // scope of NoNamesGuard
-
-  at::NoNamesGuard guard;
-  result.resize_({mat.size(0)});
+namespace at {
+namespace meta {
+  TORCH_META_FUNC(addmv)(const Tensor &self, const Tensor &mat, const Tensor &vec, const Scalar& beta, const Scalar& alpha) {
 
   Tensor self_ = self;
-  if (self.dim() == 0 || self.size(0) == 1) {
+  if (self.numel()==1 && self.dim()<=1) {
     self_ = self.expand({mat.size(0)});
   }
+  //if self was broadcastable, it's now correct size
+  //if it wasn't, it'll error out in the subsequent checks
 
   TORCH_CHECK((mat.dim() == 2 && vec.dim() == 1 && self_.dim() == 1),
-    "vector + matrix @ vector expected, got ", self_.dim(), ", ", mat.dim(), ", ", vec.dim());
-  TORCH_CHECK((mat.size(1) == vec.size(0) && mat.size(0) == self_.size(0)),
-    "size mismatch, get ", self_.size(0), ", ", mat.size(0), "x", mat.size(1), ",", vec.size(0));
+    "vector + matrix @ vector expected, got ", self.dim(), ", ", mat.dim(), ", ", vec.dim());
 
-  if (mat.numel() == 0) {
+  TORCH_CHECK((mat.size(1) == vec.size(0) && mat.size(0) == self_.size(0)),
+    "size mismatch, got ", self_.size(0), ", ", mat.size(0), "x", mat.size(1), ",", vec.size(0));
+
+  std::array<int64_t, 1> out_size{mat.sizes()[0]};
+  set_output(out_size, mat.options());
+  auto result = maybe_get_output();
+  //this check can fire for inplace op only, for all other versions result is guaranteed to be correct size
```

Comment:
note for reviewers: maybe we should update inplace structured codegen to perform this check. TensorIterator checks that inplace output is not broadcasted, so for TI-based structured kernels there would be duplication

Reply:
Hmm, I think I agree I'd prefer just checking that self and set_output have the same size.  Should be easy enough to add, and similar to the out tests we added earlier. I definitely don't like the `maybe_get_output` call here (it is in fact guaranteed to give an output because you called `set_output` earlier, and this isn't actually part of the intended public contract.)